enhancement: Add LEP 'Refactor restore for rebuild enabling' (#1629)
longhorn/longhorn: enhancements/20200721-refactor-restore-for-rebuild-enabling.md

## Motivation

### Goals
The goal of this enhancement is to simplify the restore flow so that it can work for rebuilding replicas of restore/DR volumes.

### Non-goals
This enhancement won't guarantee that the restore/DR volume activation won't be blocked by replica rebuilding.  

## Proposal
- When there is replica crashing among restore/DR volumes, new rebuilding replicas will be created as usual. But instead of following the normal replica rebuilding workflow and syncing data/files from other running replicas, the rebuilding replicas of restore/DR volumes will directly restore data from backup.
    - The normal rebuilding (file syncing) workflow implicitly considers that all existing snapshots won't change and newer data during the rebuilding will be written into volume head. But for restore/DR volumes, new data writing is directly handled by replica (sync agent server) and it will be written to underlying snapshots rather than volume heads. As a result, the normal rebuilding logic doesn't fit restore/DR volumes.  
    - In order to skip the file syncing and snapshotting and directly do restore, the rebuilding related API should be updated, which will lead to API version bumps.

> - For DR volumes, the last restored backup will be different among all replicas and some replicas may be in restore progress when the rebuilding is triggered. Hence, Longhorn manager should not specify the last backup in the restore command. In other words, engines and replicas should be intelligent enough to decide if they need to do restore and which kind of restore they need to launch.

Reviewer: > Longhorn manager should not specify the last backup in the restore command
> Longhorn manager only needs to tell the replicas what is the lastest backup they should restore.

Can you explain the difference.
Does last above refer to the previous backup?
While latest below refers to the newest backup? (in terms timeline)

Author: `the last backup` means the most recently restored backup.
`the latest backup` does means the newest backup and it hasn't been restored by the volume.

I will add more explanations for this part.

    - Actually replicas have all information about the restore status and they can decide if they need incremental restore or full restore by themself. Specifying the last backup in the restore command is redundant.

>     - Longhorn manager only needs to tell the replicas what is the lastest backup they should restore.

Reviewer 2: typo `latest` not `lastest`.

    - Longhorn manager still need to know what is the last restored backup of all replicas, since it relies on it to determine if the restore/DR volume is available/can be activated.
- Longhorn should wait for rebuild complete and check restore status before auto detachment.
    - Otherwise, the restore volume will be automatically detached when the rebuild is in progress then the rebuild is meaningless in this case.  

### User Stories

#### Replica crashes when a restore/DR volume is in restore progress
Before, the restore volume keeps state `Degraded` if there is replica crashing. And the volume will finally become `Faulted` if replicas are crashed one by one during restoring.

> After, the restore volume will start replica rebuilding automatically then be back to state `Healthy` if there is replica crashing. The volume is available as long as all replicas are not crashed at the same time. And volume will finish activation/auto-detachment after the rebuild is done.

Reviewer: Does this mean wait till state healthy and only then will the volume be available as long as there is at least one healthy replica?

Author: NO. The restore/DR volume will be `Healthy` as long as all replicas are running, which is exactly the same as normal volumes.

Previously, Longhorn won't replenish replicas, and restore/DR volumes will keep state `Degraded` if there is a replica crashing.

### User Experience In Detail

#### Replica crash on restore volume
1. Users create a restore volume and wait for restore complete.
2. When the restore is in progress, some replicas somehow get crashed. Then the volume rebuilds new replicas immediately, and it will become `Healthy` once the new replicas start rebuilding.
4. The volume will be detached automatically once the restore and the rebuild complete.

#### Replica crash on DR volume
1. Users create a DR volume.
2. Some replicas get crashed. Then the DR volume automatically rebuilds new replicas and restores the latest backup for the rebuild replicas.

> 3. Users try to activate the DR volume. The DR volume will wait for the rebuild and laest backup restore complete before detachment.

Reviewer: ```suggestion
3. Users try to activate the DR volume. The DR volume will wait for the rebuild of all replicas and successfull restoration of the latest backup before detachment.
```

7. Write data to the original volume and take the 2nd backup. (Make sure the total data size is larger than the original volume size.)
8. Wait for the 2nd backup creation complete. Then trigger DR volume last backup update by calling backup list API. 
9. Crash one random replica of the DR volume.
10. Check if the replicas will be rebuild and the restore volume can be `Healthy` after the rebuilding.
11. Wait for the expansion and restore complete. Then verify the rebuild replica is still running and the DR volume size is expanded.
12. Write data to the original volume and take the 3rd backup.
13. Wait for the 3rd backup creation complete. Then try to activate the DR volume.
14. Wait for the DR volume activated. Then launch a pod for the activated DR volume.
15. Verify the restored data of the activated DR volume.
16. Write more data to the activated DR volume. Then verify all replicas are still running.
17. Crash one random replica of the activated DR volume. Then verify a new replica will be rebuild and the volume still works fine.

### Upgrade strategy
Since there is modification for controller gRPC API, the live upgrade is not supported.

## Note

> It's possible that the restore/DR volume rebuilding somehow get stuck, or users has no time to wait for the restore/DR volume rebuilding complete. We need to provide a way that users can use the volume as soon as possible. This enhancement is trakced in https://github.com/longhorn/longhorn/issues/1512.

Reviewer 2: `tracked`
